test/includes/lxd.sh

```diff
@@ -0,0 +1,307 @@
+        check_empty_table "${daemon_dir}/lxd.db" "networks_config"
+        check_empty_table "${daemon_dir}/lxd.db" "images"
+        check_empty_table "${daemon_dir}/lxd.db" "images_aliases"
+        check_empty_table "${daemon_dir}/lxd.db" "images_properties"
+        check_empty_table "${daemon_dir}/lxd.db" "images_source"
+        check_empty_table "${daemon_dir}/lxd.db" "profiles"
+        check_empty_table "${daemon_dir}/lxd.db" "profiles_config"
+        check_empty_table "${daemon_dir}/lxd.db" "profiles_devices"
+        check_empty_table "${daemon_dir}/lxd.db" "profiles_devices_config"
+        check_empty_table "${daemon_dir}/lxd.db" "storage_pools"
+        check_empty_table "${daemon_dir}/lxd.db" "storage_pools_config"
+        check_empty_table "${daemon_dir}/lxd.db" "storage_volumes"
+        check_empty_table "${daemon_dir}/lxd.db" "storage_volumes_config"
+    fi
+
+    # teardown storage
+    "$lxd_backend"_teardown "${daemon_dir}"
+
+    # Wipe the daemon directory
+    wipe "${daemon_dir}"
+
+    # Remove the daemon from the list
+    sed "\|^${daemon_dir}|d" -i "${TEST_DIR}/daemons"
+}
+
+shutdown_lxd() {
+    # LXD_DIR is local here because since $(lxc) is actually a function, it
+    # overwrites the environment and we would lose LXD_DIR's value otherwise.
+
+    # shellcheck disable=2039
+    local LXD_DIR
+
+    daemon_dir=${1}
+    LXD_DIR=${daemon_dir}
+    daemon_pid=$(cat "${daemon_dir}/lxd.pid")
+    echo "==> Killing LXD at ${daemon_dir}"
+
+    # Kill the daemon
+    lxd shutdown || kill -9 "${daemon_pid}" 2>/dev/null || true
+}
+
+cleanup() {
+    # Allow for failures and stop tracing everything
+    set +ex
+    DEBUG=
+
+    # Allow for inspection
+    if [ -n "${LXD_INSPECT:-}" ]; then
+        if [ "${TEST_RESULT}" != "success" ]; then
+            echo "==> TEST DONE: ${TEST_CURRENT_DESCRIPTION}"
+        fi
+        echo "==> Test result: ${TEST_RESULT}"
+
+        # shellcheck disable=SC2086
+        printf "To poke around, use:\n LXD_DIR=%s LXD_CONF=%s sudo -E %s/bin/lxc COMMAND\n" "${LXD_DIR}" "${LXD_CONF}" ${GOPATH:-}
+        echo "Tests Completed (${TEST_RESULT}): hit enter to continue"
+
+        # shellcheck disable=SC2034
+        read -r nothing
+    fi
+
+    echo "==> Cleaning up"
+
+    # Kill all the LXD instances
+    # shellcheck disable=SC2039
+    local daemon_dir
+    while read -r daemon_dir; do
+        kill_lxd "${daemon_dir}"
+    done < "${TEST_DIR}/daemons"
+
+    # Cleanup leftover networks
+    # shellcheck disable=SC2009
+    ps aux | grep "interface=lxdt$$ " | grep -v grep | awk '{print $2}' | while read -r line; do
+        kill -9 "${line}"
+    done
+    if [ -e "/sys/class/net/lxdt$$" ]; then
+        ip link del lxdt$$
+    fi
+
+    # Wipe the test environment
+    wipe "${TEST_DIR}"
+
+    echo ""
+    echo ""
+    if [ "${TEST_RESULT}" != "success" ]; then
+        echo "==> TEST DONE: ${TEST_CURRENT_DESCRIPTION}"
+    fi
+    echo "==> Test result: ${TEST_RESULT}"
+}
```

Comment:
I think cleanup() should remain in main.sh as it's very specific to this particular testsuite and also makes it pretty hard to figure out how main.sh even works when the last thing it does is hidden away.